File: superleague-lua/superleague.lua
#!/usr/bin/env lua

local json = require("dkjson")

function new_league_row(team_name)
    return {team = team_name, wins = 0, ties = 0, defeats = 0, goals_scored = 0, goals_conceded = 0}
end

function sort(a, b)
    if a["points"] > b["points"] then
        return true
    elseif a["points"] < b["points"] then
        return false
    elseif a["goals_difference"] > b["goals_difference"] then
        return true
    elseif a["goals_difference"] < b["goals_difference"] then
        return false
    elseif a["wins"] > b["wins"] then
        return true
    elseif a["wins"] < b["wins"] then
        return false
    elseif a["team"] < b["team"] then
        return true
    elseif a["team"] > b["team"] then
        return false
    end
end

matches_path = arg[1]
local matches_file = io.open(matches_path, "r")
local matches_data = matches_file:read("*a")

local matches = json.decode(matches_data)
local league = {}

for _, match in ipairs(matches) do
    home_team = match["homeTeam"]
    away_team = match["awayTeam"]
    home_goals = match["homeGoals"]
    away_goals = match["awayGoals"]

    if league[home_team] == nil then
        league[home_team] = new_league_row(home_team)
    end
    if league[away_team] == nil then
        league[away_team] = new_league_row(away_team)
    end

    league[home_team]["goals_scored"] = league[home_team]["goals_scored"] + home_goals
    league[home_team]["goals_conceded"] = league[home_team]["goals_conceded"] + away_goals
    league[away_team]["goals_scored"] = league[away_team]["goals_scored"] + away_goals
    league[away_team]["goals_conceded"] = league[away_team]["goals_conceded"] + home_goals 

    if home_goals > away_goals then
        league[home_team]["wins"] = league[home_team]["wins"] + 1
        league[away_team]["defeats"] = league[away_team]["defeats"] + 1
    elseif home_goals < away_goals then
        league[home_team]["defeats"] = league[home_team]["defeats"] + 1
        league[away_team]["wins"] = league[away_team]["wins"] + 1
    else
        league[home_team]["ties"] = league[home_team]["ties"] + 1
        league[away_team]["ties"] = league[away_team]["ties"] + 1
    end
end

for team, row in pairs(league) do
    row["goals_difference"] = row["goals_scored"] - row["goals_conceded"]
    row["points"] = row["wins"] * 3 + row["ties"]
end

ordered = {}
for _, row in pairs(league) do
    table.insert(ordered, row)
end
table.sort(ordered, sort)

title = string.format("%25s %2s %2s %2s %2s %2s %2s %3s %2s",
                      "Name", "#", "w", "d", "l", "+", "-", "=", "P")
print(title)
print(string.rep("-", string.len(title)))
for i, row in pairs(ordered) do
    print(string.format("%25s %2d %2d %2d %2d %2d %2d %3d %2d",
                        row["team"], i+1,
                        row["wins"], row["ties"], row["defeats"],
                        row["goals_scored"], row["goals_conceded"], row["goals_difference"],
                        row["points"]))
end

File: superleague-php/superleague.php
<?php

$rows = array();
$json_data = file_get_contents($argv[1]);
$season = json_decode($json_data, true);

function newEntry($teamName) {
    return array(
        "team" => $teamName,
        "rank" => 0,
        "wins" => 0,
        "ties" => 0,
        "defeats" => 0,
        "goalsScored" => 0,
        "goalsConceded" => 0,
        "goalsDifference" => 0,
        "points" => 0,
    );
}

foreach ($season as $match) {
    $homeTeam = $match["homeTeam"];
    $awayTeam = $match["awayTeam"];
    $homeGoals = $match["homeGoals"];
    $awayGoals = $match["awayGoals"];

    if (!array_key_exists($homeTeam, $rows)) {
        $rows[$homeTeam] = newEntry($homeTeam);
    }
    if (!array_key_exists($awayTeam, $rows)) {
        $rows[$awayTeam] = newEntry($awayTeam);
    }

    $rows[$homeTeam]["goalsScored"] += $homeGoals;
    $rows[$homeTeam]["goalsConceded"] += $awayGoals;
    $rows[$awayTeam]["goalsScored"] += $awayGoals;
    $rows[$awayTeam]["goalsConceded"] += $homeGoals;

    if ($homeGoals > $awayGoals) {
        $rows[$homeTeam]["wins"]++;
        $rows[$awayTeam]["defeats"]++;
    } elseif ($homeGoals < $awayGoals) {
        $rows[$awayTeam]["wins"]++;
        $rows[$homeTeam]["defeats"]++;
    } else {
        $rows[$homeTeam]["ties"]++;
        $rows[$homeTeam]["points"]++;
        $rows[$awayTeam]["ties"]++;
    }
}

$rows = array_map(function ($row) {
	$row["goalsDifference"] = $row["goalsScored"] - $row["goalsConceded"];
	$row["points"] = $row["wins"] * 3 + $row["ties"];
	return $row;
}, $rows);

usort($rows, function ($a, $b) {
    if ($a["points"] > $b["points"]) {
        return -1;
    } elseif ($a["points"] < $b["points"]) {
        return 1;
    } elseif ($a["goalsDifference"] > $b["goalsDifference"]) {
        return -1;
    } elseif ($a["goalsDifference"] < $b["goalsDifference"]) {
        return 1;
    } elseif ($a["wins"] < $b["wins"]) {
        return -1;
    } elseif ($a["wins"] < $b["wins"]) {
        return 1;
    } elseif (strcasecmp($a["team"], $b["team"]) < 0) {
        return -1;
    } elseif (strcasecmp($a["team"], $b["team"]) > 0) {
        return 1;
    } else {
        return 0;
    }
});

print("                     Name  #  w  d  l  +  -   =  P\n");
print("--------------------------------------------------\n");
foreach ($rows as $k => $l) {
    printf("%25s %2d %2d %2d %2d %2d %2d %3d %2d\n",
        $l["team"], $k+1, $l["wins"], $l["ties"], $l["defeats"],
        $l["goalsScored"], $l["goalsConceded"], $l["goalsDifference"], $l["points"]);
}

?>
